runtime/service.go

```diff
@@ -1178,24 +1178,21 @@ func (s *service) Shutdown(requestCtx context.Context, req *taskAPI.ShutdownRequ
 	_, err := s.agentClient.Shutdown(requestCtx, req)
 	if err != nil {
 		shutdownErr = multierror.Append(shutdownErr, errors.Wrap(err, "failed to shutdown VM Agent"))
-	}
 
-	err = s.machine.StopVMM()
-	if err != nil {
-		shutdownErr = multierror.Append(shutdownErr, errors.Wrap(err, "failed to gracefully stop VM"))
+		if err := s.machine.StopVMM(); err != nil {
+			shutdownErr = multierror.Append(shutdownErr, errors.Wrap(err, "failed to shutdown VMM"))
+		}
 	}
 
-	err = os.RemoveAll(s.shimDir.RootPath())
-	if err != nil {
-		shutdownErr = multierror.Append(shutdownErr, errors.Wrapf(err, "failed to remove VM dir %q during shutdown", s.shimDir.RootPath()))
+	if err := s.machine.Wait(context.Background()); err != nil {
+		shutdownErr = multierror.Append(shutdownErr, err)
```

Comment:
I'm little concerned that we are hardcoding an assumption about the rootfs (that agent exiting results in the machine exiting) which is true for our image-builder's image but not for all images that could be used by any user in general (they are free to use their own too).

I think we could address this by having a timeout on the call to `machine.Wait` that was added here (so then we don't get into situations in which `Shutdown` blocks indefinitely). However, that would create some hard-to-follow interactions between here and `StopVM`, which already has a timeout and calls `Shutdown` in a goroutine.

Overall, I'm wondering if it would make sense to migrate the timeout logic present in `StopVM` (setup on lines 555-558) to be enforced within `Shutdown` instead. Because `StopVM` calls `Shutdown`, we could then still end up with the correct timeout logic but without having confusing multiple layers of timeouts being enforced. Let me know what you think.

Reply:
My thoughts were "Shutdown for graceful shutdown. StopVM for forcefully shutdown", but I agree that the relationship between StopVM and Shutdown is hard to follow. Will move the timeout functionality to Shutdown().

BTW, do you know the reason we expose Shutdown() as a public method (instead of being "shutdown()")?

Reply:
> BTW, do you know the reason we expose Shutdown() as a public method (instead of being "shutdown()")?

Yeah, it's because it's actually part of the API we need to implement as a containerd shim: https://github.com/containerd/containerd/blob/master/runtime/v2/task/shim.proto#L35

Reply:
Thanks! And my comment was wrong. Shutdown is also calling StopVMM...

Reply:
Addressed in ddbde81ea22749cba5a1ba058049ee0afd5673e1